### PR TITLE
Adding different entry point options to Train Job API

### LINKED ISSUE #203: Why doesn't PredictionEndpoint/Task interface support the append builder?
It looks like the PredictionEndpoint interface doesn't support the append builder.

It doesn't look like there's a way to pass to get_builder `needs_deps_installation=False` to force it to use the append builder rather than the ClusterBuilder (Kaniko)

https://github.com/kubeflow/fairing/blob/642530567d0255f88b0f83b0a40b3325c303bfa2/fairing/ml_tasks/tasks.py#L36

/cc @karthikv2k

### PR DESCRIPTION
**Pull request by karthikv2k**
**What this PR does / why we need it**:
1. Currently TrainJob API only supports Python function/class. This PR adds support for executing python files and full notebook execution.
2. Exposes pod_spec_mutator argument in TrainJob API that lets user to customize pod spec like CPUs and mount volumes
3. Improved how we pick the builder in TrainJob API. Now it uses AppendBuilder when no python package installation is required. Also makes sure right gcp secret is available before executing cluster builder.
4. Improved error messages
5. Added an example notebook - examples/train_job_api/main.ipynb

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #203 #166 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.
NA

**Release note**:

```release-note
TrainJob API supports python file and full notebook execution.
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/fairing/279)

### CODE REVIEW

File: fairing/backends/backends.py [hunk 1]
```diff
@@ -109,7 +146,11 @@ def get_builder(self, preprocessor, base_image, registry, needs_deps_installatio
         pod_spec_mutators.append(gcp.add_gcp_credentials_if_exists)
         # TODO (karthikv2k): Add cloud build as the deafult
         # once https://github.com/kubeflow/fairing/issues/145 is fixed
-        if fairing.utils.is_running_in_k8s():
+        if not needs_deps_installation:
```
Comment (abhi-g) [inline fairing/backends/backends.py]: This logic for deciding which Builder to use would be good to factor into a separate function that returns an Enum or equivalent in python. that would make it easier to keep in sync at multiple invocations.

**Comment by abhi-g** [comment]
I like the template you have going here for the description of the issue! we should discuss this in one of the community meetings.

**Comment by karthikv2k** [comment]
> I like the template you have going here for the description of the issue! we should discuss this in one of the community meetings.

credit goes to @gyliu513 #270

File: fairing/backends/backends.py (hunk 1 above)
Comment (karthikv2k) [reply fairing/backends/backends.py]: I made the changes but let me do that as a separate PR to keep the changes to minimum.

**Pull request merged**